public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin2.tsx

```diff
@@ -0,0 +1,199 @@
+import { useLayoutEffect, useMemo, useRef } from 'react';
+import uPlot from 'uplot';
+
+import { DataFrame, FieldType } from '@grafana/data';
+import { alpha } from '@grafana/data/src/themes/colorManipulator';
+import { UPlotConfigBuilder } from '@grafana/ui';
+
+import Flatbush from '../../barchart/flatbush.esm';
+
+interface ExemplarsPlugin2Props {
+  config: UPlotConfigBuilder;
+  exemplars: DataFrame;
+  matchByField: string;
+  seriesLabels: string[];
+  onHover?: (exemplarIdx: number, x: number, y: number) => void;
+  onLeave?: (exemplarIdx: number) => void;
+  // timeZone: TimeZone;
+  // getFieldLinks: (field: Field, rowIndex: number) => Array<LinkModel<Field>>;
+}
+
+// diamonds!
+let ptSizeCss = 7; // in logical pixels
+// let halfDiagMult = Math.sqrt(2) / 2;
+
+// TODO: writing this shape to a raster sprite and then drawImage/imageSmoothingEnabled
+// is probably faster than looped moveTo/lineTo/fill commands (for complex shapes)
+function drawMark(ctx: CanvasRenderingContext2D, x: number, y: number, size = 7) {
+  // let halfDiag = size * halfDiagMult;
+  let halfDiag = size / 2;
+
+  ctx.beginPath();
+  ctx.moveTo(x - halfDiag, y);
+  ctx.lineTo(x, y - halfDiag);
+  ctx.lineTo(x + halfDiag, y);
+  ctx.lineTo(x, y + halfDiag);
+  ctx.fill();
+}
+
+export const ExemplarsPlugin2 = ({ config, exemplars, matchByField, seriesLabels }: ExemplarsPlugin2Props) => {
+  let exemplarsRef = useRef<DataFrame>(exemplars);
+
+  exemplarsRef.current = exemplars;
+
+  // filter out exemplars of hidden series, group by color to minimize ctx.fillStyle mutation
+  let exemplarIdxsByColor = useMemo(() => {
+    let seriesLabelsToIdx = new Map();
+    seriesLabels.forEach((l, i) => seriesLabelsToIdx.set(l, i));
+
+    let exemplarLabels = exemplars.fields.find((f) => f.name === matchByField)!.values.toArray();
```

Comment:
this is what extracts the exemplar field that will be matched up with `series.name`, for colors / visibility.